bad request body causes failures

If `json.loads` can parse the body of an `application/json` request, but it returns something other than a `dict`, both Flask and Django integrations fail to process the request.

For example, a request body of `1` is parseable, but causes the Flask integration raise the exception
```
...
/usr/local/lib/python3.10/site-packages/appmap/_implementation/web_framework.py:164: in before_request_hook
    start, call_event_id = self.before_request_main(rec, request)
/usr/local/lib/python3.10/site-packages/appmap/flask.py:99: in before_request_main
    message_parameters=request_params(request),
/usr/local/lib/python3.10/site-packages/appmap/flask.py:39: in request_params
    params.update(req.json or {})
/usr/local/lib/python3.10/site-packages/werkzeug/datastructures.py:552: in update
    for key, value in iter_multi_items(mapping):
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

mapping = 1

    def iter_multi_items(mapping):
        """Iterates over the items of a mapping yielding keys and values
        without dropping any from more complex structures.
        """
        if isinstance(mapping, MultiDict):
            yield from mapping.items(multi=True)
        elif isinstance(mapping, dict):
            for key, value in mapping.items():
                if isinstance(value, (tuple, list)):
                    for v in value:
                        yield key, v
                else:
                    yield key, value
        else:
>           yield from mapping
E           TypeError: 'int' object is not iterable

/usr/local/lib/python3.10/site-packages/werkzeug/datastructures.py:35: TypeError
```

--- a/_appmap/configuration.py
+++ b/_appmap/configuration.py
@@ -4,7 +4,7 @@ Manage Configuration AppMap recorder for Python.
 
 import inspect
 import logging
-import re
+import os
 import sys
 from functools import lru_cache
 from os.path import realpath
@@ -23,12 +23,6 @@ from .importer import Filter, Importer
 from .instrument import instrument
 
 logger = logging.getLogger(__name__)
-
-
-def warn_config_missing(path):
-    """Display a warning about missing config file in path."""
-    name = path.resolve().parent.name
-    package = re.sub(r"\W", ".", name).lower()
 
 
 def default_app_name(rootdir):
@@ -99,7 +93,6 @@ def find_top_packages(rootdir):
     # Use a set so we don't get duplicates, e.g. if the project's
     # build process copies its source to a subdirectory.
     packages = set()
-    import os
 
     def excluded(d):
         excluded = d == "node_modules" or d[0] == "."
@@ -219,7 +212,7 @@ class Config:
             Env.current.enabled = False
             self.file_valid = False
             try:
-                self._config = yaml.safe_load(path.read_text())
+                self._config = yaml.safe_load(path.read_text(encoding="utf-8"))
                 if not self._config:
                     # It parsed, but was (effectively) empty.
                     self._config = self.default
@@ -264,7 +257,7 @@ It will be created with this configuration:
         basedir = filepath.parent
         if not basedir.exists():
             basedir.mkdir(parents=True, exist_ok=True)
-        with open(filepath, "w") as f:
+        with open(filepath, "w", encoding="utf-8") as f:
             f.write(yaml.dump(config, sort_keys=True))
 
 
@@ -393,7 +386,7 @@ class BuiltinFilter(MatcherFilter):
     def __init__(self, *args, **kwargs):
         matchers = []
         if Env.current.enabled:
-            matchers = [PathMatcher(f) for f in {"os.read", "os.write"}]
+            matchers = [PathMatcher(f) for f in ("os.read", "os.write")]
         super().__init__(matchers, *args, **kwargs)
 
 

--- a/_appmap/generation.py
+++ b/_appmap/generation.py
@@ -106,9 +106,9 @@ class AppMapEncoder(json.JSONEncoder):
     def default(self, o):
         if isinstance(o, Event):
             return o.to_dict()
-        elif isinstance(o, ClassMapDict):
+        if isinstance(o, ClassMapDict):
             return list(o.values())
-        elif isinstance(o, ClassMapEntry):
+        if isinstance(o, ClassMapEntry):
             return o.to_dict()
 
         return json.JSONEncoder.default(self, o)

--- a/_appmap/test/data/django/app/__init__.py
+++ b/_appmap/test/data/django/app/__init__.py
@@ -13,6 +13,7 @@ django.conf.settings.configure(
             "DIRS": [Path(__file__).parent],
         }
     ],
+    SECRET_KEY="not-a-secret",
 )
 
 django.setup()

--- a/_appmap/test/test_django.py
+++ b/_appmap/test/test_django.py
@@ -7,7 +7,6 @@ from pathlib import Path
 from threading import Thread
 
 import django
-import django.conf
 import django.core.handlers.exception
 import django.db
 import django.http
@@ -25,13 +24,14 @@ from ..test.helpers import DictIncluding
 
 # Make sure assertions in web_framework get rewritten (e.g. to show
 # diffs in generated appmaps)
-pytest.register_assert_rewrite("test.web_framework")
+pytest.register_assert_rewrite("_appmap.test.web_framework")
 
-from .web_framework import (  # pylint: disable=wrong-import-position
-    _TestRecordRequests,
-    exec_cmd,
-    wait_until_port_is,
-)
+# pylint: disable=unused-import,wrong-import-position
+from .web_framework import TestRemoteRecording  # pyright:ignore
+from .web_framework import TestRequestCapture  # pyright: ignore
+from .web_framework import _TestRecordRequests, exec_cmd, wait_until_port_is
+
+# pylint: enable=unused-import
 
 sys.path += [str(Path(__file__).parent / "data" / "django")]
 
@@ -120,7 +120,9 @@ class ClientAdaptor(django.test.Client):
 
 
 @pytest.fixture
-def client():
+def client(settings):
+    # We might be able to set DEBUG in the app's settings. But, using the settings fixture here
+    # ensures that all the settings (esp MIDDLEWARE) will be reset before each test run.
     settings.DEBUG = True
     return ClientAdaptor()
 

--- a/_appmap/test/test_flask.py
+++ b/_appmap/test/test_flask.py
@@ -6,22 +6,23 @@ from threading import Thread
 
 import flask
 import pytest
-import requests
 
 from _appmap.env import Env
 from _appmap.metadata import Metadata
-from appmap.flask import AppmapFlask
+from appmap.flask import AppmapFlask, FlaskInserter
 
 from ..test.helpers import DictIncluding
-from .web_framework import (  # pylint: disable=unused-import
-    _TestRecordRequests,
-    exec_cmd,
-    wait_until_port_is,
-)
 
 # Make sure assertions in web_framework get rewritten (e.g. to show
 # diffs in generated appmaps)
-pytest.register_assert_rewrite("test.web_framework")
+pytest.register_assert_rewrite("_appmap.test.web_framework")
+
+# pylint: disable=unused-import,wrong-import-position
+from .web_framework import TestRemoteRecording  # pyright:ignore
+from .web_framework import TestRequestCapture  # pyright: ignore
+from .web_framework import _TestRecordRequests, exec_cmd, wait_until_port_is
+
+# pylint: enable=unused-import
 
 
 @pytest.fixture(name="app")
@@ -29,6 +30,8 @@ def flask_app(data_dir, monkeypatch):
     monkeypatch.syspath_prepend(data_dir / "flask")
 
     Env.current.set("APPMAP_CONFIG", data_dir / "flask" / "appmap.yml")
+
+    monkeypatch.setenv("FLASK_DEBUG", "True")
 
     import app  # pyright: ignore pylint: disable=import-error,import-outside-toplevel
 

--- a/_appmap/test/web_framework.py
+++ b/_appmap/test/web_framework.py
@@ -160,10 +160,12 @@ class TestRequestCapture:
             )
         ]
 
-    @staticmethod
-    def test_post_bad_json(events, client):
+    @pytest.mark.parametrize("bad_json", ["bad json", 1, False, None, [2, 3]])
+    def test_post_bad_json(self, events, client, bad_json):
         client.post(
-            "/test?my_param=example", data="bad json", content_type="application/json"
+            "/test?my_param=example",
+            data=str(bad_json),
+            content_type="application/json",
         )
 
         assert events[0].message == [
@@ -213,7 +215,6 @@ class TestRequestCapture:
         assert np == expected
 
     @staticmethod
-    @pytest.mark.appmap_enabled
     def test_message_path_segments(events, client):
         client.get("/post/alice/42/summary")
 
@@ -225,14 +226,14 @@ class TestRequestCapture:
         ]
 
 
-class TestRecording:
+@pytest.mark.appmap_enabled(env={"APPMAP_RECORD_REQUESTS": "false"})
+class TestRemoteRecording:
     """Common tests for remote recording."""
 
     @staticmethod
-    def test_appmap_disabled(client, monkeypatch):
-        # since APPMAP records by default, disable it explicitly
-        monkeypatch.setenv("APPMAP", "false")
-        appmap.initialize()  # pylint: disable=protected-access
+    # since APPMAP records by default, disable it explicitly
+    @pytest.mark.appmap_enabled(appmap_enabled=False)
+    def test_appmap_disabled(client):
         assert not appmap.enabled()
         assert not Env.current.enables("remote")
 
@@ -240,7 +241,6 @@ class TestRecording:
         assert res.status_code == 404
 
     @staticmethod
-    @pytest.mark.appmap_enabled
     def test_starts_disabled(client):
         res = client.get("/_appmap/record")
         assert res.status_code == 200
@@ -253,13 +253,11 @@ class TestRecording:
         assert data == {"enabled": False}
 
     @staticmethod
-    @pytest.mark.appmap_enabled
     def test_can_be_enabled(client):
         res = client.post("/_appmap/record")
         assert res.status_code == 200
 
     @staticmethod
-    @pytest.mark.appmap_enabled
     def test_can_only_enable_once(client):
         res = client.post("/_appmap/record")
         assert res.status_code == 200
@@ -267,7 +265,6 @@ class TestRecording:
         assert res.status_code == 409
 
     @staticmethod
-    @pytest.mark.appmap_enabled
     def test_can_record(data_dir, client):
         res = client.post("/_appmap/record")
         assert res.status_code == 200
@@ -415,23 +412,23 @@ class _TestRecordRequests:
                 )
 
                 with open(appmap_file_name, encoding="utf-8") as f:
-                    appmap = json.loads(f.read())
+                    recording = json.loads(f.read())
 
                     # Every event should come from the same thread
                     thread_ids = {
-                        event["thread_id"]: True for event in appmap["events"]
+                        event["thread_id"]: True for event in recording["events"]
                     }
                     assert len(thread_ids) == 1
 
                     # AppMap should contain only one request and response
                     http_server_requests = [
                         event["http_server_request"]
-                        for event in appmap["events"]
+                        for event in recording["events"]
                         if "http_server_request" in event
                     ]
                     http_server_responses = [
                         event["http_server_response"]
-                        for event in appmap["events"]
+                        for event in recording["events"]
                         if "http_server_response" in event
                     ]
                     assert len(http_server_requests) == 1

--- a/_appmap/web_framework.py
+++ b/_appmap/web_framework.py
@@ -168,7 +168,7 @@ class AppmapMiddleware(ABC):
         return rec, start, call_event_id
 
     @abstractmethod
-    def before_request_main(self, rec):
+    def before_request_main(self, rec, req):
         """Specify the main operations to be performed by a request is processed."""
 
     def after_request_hook(

--- a/_appmap/web_framework.py
+++ b/_appmap/web_framework.py
@@ -11,6 +11,7 @@ import time
 from abc import ABC, abstractmethod
 from contextvars import ContextVar
 from hashlib import sha256
+from json.decoder import JSONDecodeError
 from tempfile import NamedTemporaryFile
 
 from . import generation, remote_recording
@@ -21,6 +22,10 @@ from .utils import root_relative_path, scenario_filename
 
 logger = logging.getLogger(__name__)
 request_recorder = ContextVar("appmap_request_recorder")
+
+# These are the errors that can get raised when trying to update params based on the results of
+# parsing the body of an application/json request:
+JSON_ERRORS = (JSONDecodeError, AttributeError, TypeError, ValueError)
 
 
 class TemplateEvent(Event):  # pylint: disable=too-few-public-methods

--- a/appmap/command/appmap_agent_status.py
+++ b/appmap/command/appmap_agent_status.py
@@ -44,7 +44,7 @@ def discover_pytest_tests():
     #
     collector = AgentFileCollector()
 
-    import pytest
+    import pytest  # pylint: disable=import-outside-toplevel
 
     pytest.main(
         [

--- a/appmap/django.py
+++ b/appmap/django.py
@@ -32,7 +32,7 @@ from _appmap.instrument import is_instrumentation_disabled
 from _appmap.metadata import Metadata
 from _appmap.recorder import Recorder
 from _appmap.utils import patch_class, values_dict
-from _appmap.web_framework import AppmapMiddleware, MiddlewareInserter
+from _appmap.web_framework import JSON_ERRORS, AppmapMiddleware, MiddlewareInserter
 from _appmap.web_framework import TemplateHandler as BaseTemplateHandler
 
 logger = logging.getLogger(__name__)
@@ -151,7 +151,7 @@ def request_params(request):
             # directly reading request. This way the application can still
             # access the body which is now cached in the request object.
             params.update(json.loads(request.body))
-        except (json.decoder.JSONDecodeError, AttributeError):
+        except JSON_ERRORS:
             pass  # invalid json or not an object
 
     return values_dict(params.lists())

--- a/appmap/django.py
+++ b/appmap/django.py
@@ -66,7 +66,10 @@ class ExecuteWrapper:  # pylint: disable=too-few-public-methods
     def __init__(self):
         self.recorder = Recorder.get_current()
 
-    def __call__(self, execute, sql, params, many, context):
+    # This signature is correct, the implementation confuses pylint:
+    def __call__(
+        self, execute, sql, params, many, context
+    ):  # pylint: disable=too-many-arguments
         start = time.monotonic()
         try:
             return execute(sql, params, many, context)
@@ -112,7 +115,8 @@ def wrapped_execute(self, sql, params=None):
     disabled, to avoid capturing queries not issued by the application."""
     if is_instrumentation_disabled():
         # fmt: off
-        return super().execute(sql, params) # pyright: ignore (seems like [superCallZeroArgForm] should suppress this, it doesn't)
+        # Seems like pyright: ignore [superCallZeroArgForm] should suppress this, it doesn't:
+        return super().execute(sql, params) # pyright: ignore
         # fmt on
     return original_execute(self, sql, params)
 
@@ -257,26 +261,26 @@ class Middleware(AppmapMiddleware):
 
         return response
 
-    def before_request_main(self, rec, request):
+    def before_request_main(self, rec, req):
         add_metadata()
         start = time.monotonic()
-        params = request_params(request)
+        params = request_params(req)
         try:
-            resolved = resolve(request.path_info)
+            resolved = resolve(req.path_info)
             params.update(resolved.kwargs)
-            normalized_path_info = normalize_path_info(request.path_info, resolved)
+            normalized_path_info = normalize_path_info(req.path_info, resolved)
         except Resolver404:
             # If the request was for a bad path (e.g. when an app
             # is testing 404 handling), resolving will fail.
             normalized_path_info = None
 
         call_event = HttpServerRequestEvent(
-            request_method=request.method,
-            path_info=request.path_info,
+            request_method=req.method,
+            path_info=req.path_info,
             message_parameters=params,
             normalized_path_info=normalized_path_info,
-            protocol=request.META["SERVER_PROTOCOL"],
-            headers=request.headers,
+            protocol=req.META["SERVER_PROTOCOL"],
+            headers=req.headers,
         )
         rec.add_event(call_event)
 
@@ -312,13 +316,16 @@ class DjangoInserter(MiddlewareInserter):
         stack[0:0] = new_middleware
         # Django is ok with settings.MIDDLEWARE being any kind iterable. Update it, without changing
         # its type, if we can.
+        msg = (
+            "Don't know how to update settings.MIDDLEWARE of type %s, recording is not enabled.",
+        )
         if isinstance(self.settings.MIDDLEWARE, list):
             self.settings.MIDDLEWARE = stack
         elif isinstance(self.settings.MIDDLEWARE, tuple):
             self.settings.MIDDLEWARE = tuple(stack)
         else:
             logger.warning(
-                "Don't know how to update settings.MIDDLEWARE of type %s, recording is not enabled.",
+                msg,
                 type(self.settings.MIDDLEWARE),
             )
 

--- a/appmap/flask.py
+++ b/appmap/flask.py
@@ -38,8 +38,13 @@ def request_params(req):
     params.update(req.view_args or {})
     try:
         params.update(req.json or {})
-    except BadRequest:
-        pass  # probably a JSON parse error
+    # pylint is wrong about this "exception operation":
+    except (BadRequest,) + JSON_ERRORS:  # pylint: disable=wrong-exception-operation
+        # If request claims to be application/json, but its body is unparsable, BadRequest will be
+        # raised. json is pretty forgiving about what it will parse, though, so req.json may not be
+        # a dict. When that's the case, ValueError or TypeError may be raised when trying to update
+        # params.
+        pass
 
     return values_dict(params.lists())
 
@@ -169,7 +174,7 @@ class FlaskInserter(MiddlewareInserter):
         return getattr(self.app, _REMOTE_ENABLED_ATTR, None)
 
 
-def install_extension(wrapped, instance, args, kwargs):
+def install_extension(wrapped, _, args, kwargs):
     app = wrapped(*args, **kwargs)
     if app:
         FlaskInserter(app).run()

--- a/appmap/flask.py
+++ b/appmap/flask.py
@@ -15,7 +15,7 @@ from _appmap.event import HttpServerRequestEvent, HttpServerResponseEvent
 from _appmap.flask import app as remote_recording_app
 from _appmap.metadata import Metadata
 from _appmap.utils import patch_class, values_dict
-from _appmap.web_framework import AppmapMiddleware, MiddlewareInserter
+from _appmap.web_framework import JSON_ERRORS, AppmapMiddleware, MiddlewareInserter
 from _appmap.web_framework import TemplateHandler as BaseTemplateHandler
 from appmap import wrapt
 

--- a/appmap/pytest.py
+++ b/appmap/pytest.py
@@ -54,8 +54,10 @@ if not Env.current.is_appmap_repo and Env.current.enables("pytest"):
         # TestCase run by pytest.)
         if hasattr(item, "_testcase"):
             setattr(
-                item._testcase, "_appmap_pytest_recording", True
-            )  # pylint: disable=protected-access
+                item._testcase,  # pylint: disable=protected-access
+                "_appmap_pytest_recording",
+                True,
+            )
             item.obj = recorded_testcase(item)(item.obj)
 
     @pytest.hookimpl(hookwrapper=True)

--- a/appmap/unittest.py
+++ b/appmap/unittest.py
@@ -4,6 +4,7 @@ from contextlib import contextmanager
 
 from _appmap import testing_framework
 from _appmap.env import Env
+from _appmap.utils import get_function_location
 from appmap import wrapt
 
 logger = logging.getLogger(__name__)
@@ -13,7 +14,6 @@ def setup_unittest():
     session = testing_framework.session("unittest", "tests")
 
     def get_test_location(cls, method_name):
-        from _appmap.utils import get_function_location
 
         fn = getattr(cls, method_name)
         return get_function_location(fn)

--- a/pylintrc
+++ b/pylintrc
@@ -6,8 +6,13 @@ ignore-paths= ^_appmap/test/data
 # run arbitrary code.
 extension-pkg-whitelist=
 
+# Return non-zero exit code if any of these messages/categories are detected,
+# even if score is above --fail-under value. Syntax same as enable. Messages
+# specified are enabled, while categories only check already-enabled messages.
+fail-on=forgotten-debug-statement
+
 # Specify a score threshold to be exceeded before program exits with error.
-fail-under=9.45
+fail-under=9.52
 
 # Add files or directories to the blacklist. They should be base names, not
 # paths.
@@ -61,7 +66,7 @@ confidence=
 # --enable=similarities". If you want to run only the classes checker, but have
 # no Warning level messages displayed, use "--disable=all --enable=classes
 # --disable=W".
-disable=consider-using-fstring, invalid-name, missing-function-docstring, missing-class-docstring, missing-module-docstring, consider-using-f-string
+disable=invalid-name, missing-function-docstring, missing-class-docstring, missing-module-docstring, consider-using-f-string
 
 # Enable the message, report, category or checker with the given id(s). You can
 # either give multiple identifier separated by comma (,) or put this option

